include/zephyr/drivers/retention.h

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for data retention drivers
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_RETENTION_
+#define ZEPHYR_INCLUDE_DRIVERS_RETENTION_
+
+#include <stdint.h>
+#include <stddef.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Data Retention Driver Interface
+ * @defgroup retention_interface Data Retention Driver Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+/**
+ * @typedef	retention_size_api
+ * @brief	Callback API to get size of retained data area.
+ * See retention_size() for argument description
+ */
+typedef int32_t (*retention_size_api)(const struct device *dev);
+
+/**
+ * @typedef	retention_is_valid_api
+ * @brief	Callback API to check if retained data is valid (correct header and checksum,
+ *		if enabled).
+ * See retention_is_valid() for argument description
+ */
+typedef int32_t (*retention_is_valid_api)(const struct device *dev);
+
+/**
+ * @typedef	retention_read_api
+ * @brief	Callback API to read retained data.
+ * See retention_read() for argument description
+ */
+typedef int32_t (*retention_read_api)(const struct device *dev, size_t offset, uint8_t *buffer,
+				      size_t size);
+
+/**
+ * @typedef	retention_write_api
+ * @brief	Callback API to write retained data (will set valid bit if supported by driver).
+ * See retention_write() for argument description
+ */
+typedef int32_t (*retention_write_api)(const struct device *dev, size_t offset, uint8_t *buffer,
+				       size_t size);
+
+/**
+ * @brief Data retention driver API
+ * API which can be used by a device to retain data when a device reboots or powers down
+ */
+struct retention_driver_api {
+	retention_size_api size;
+	retention_is_valid_api is_valid;
+	retention_read_api read;
+	retention_write_api write;
+};
+
+/**
+ * @brief		Returns the size of the retention area.
+ *
+ * @param dev		Retention device (area) to use.
+ *
+ * @retval		Positive value indicating size in bytes on success, else negative errno
+ * 			code.
+ */
+__syscall int32_t retention_size(const struct device *dev);
+
+static inline int32_t z_impl_retention_size(const struct device *dev)
+{
+	struct retention_driver_api *api = (struct retention_driver_api *)dev->api;
+
+	if (!api->size) {
+		return -ENOSYS;
+	}
+
+	return api->size(dev);
+}
+
+/**
+ * @brief		Checks if the underlying data in the retention area is valid, if the
+ * 			hardware supports it.
+ *
+ * @param dev		Retention device (area) to use.
+ *
+ * @retval		1 on success when data is valid, 0 when data is not valid, else negative
+ *			errno code.
```

Comment:
What should callback return when the hardware does not implement the feature?

Reply:
This is defined in the design guidelines https://docs.zephyrproject.org/latest/develop/api/design_guidelines.html#return-codes

```
    APIs that are supported but not implemented shall return -ENOSYS.

    Optional APIs that are not supported by the hardware should be implemented and the return code in this case shall be -ENOTSUP.
```